runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -70,46 +72,79 @@ static inline void* alignHelper(void* base_ptr, size_t offset, size_t alignment)
 
   void* cur_chunk_base = NULL;
-  size_t cur_heap_size;
 
-  // this function can be called concurrently and it looks like jemalloc
-  // doesn't call it inside a lock, so we need to protect it ourselves
-  pthread_mutex_lock(&heap.alloc_lock);
+  if (heap.isFixed) {
+    //
+    // Get more space out of the fixed heap.
+    //
+    size_t cur_heap_size;
+
+    // this function can be called concurrently and it looks like jemalloc
+    // doesn't call it inside a lock, so we need to protect it ourselves
+    pthread_mutex_lock(&heap.alloc_lock);
+
+    // compute our current aligned pointer into the shared heap
+    //
+    //   jemalloc 4.5.0 man: "The alignment parameter is always a power of two
+    //   at least as large as the chunk size."
+    cur_chunk_base = alignHelper(heap.base, heap.cur_offset, alignment);
+
+    // jemalloc 4.5.0 man: "If chunk is not NULL, the returned pointer must be
+    // chunk on success or NULL on error"
+    if (chunk && chunk != cur_chunk_base) {
+      pthread_mutex_unlock(&heap.alloc_lock);
+      return NULL;
+    }
 
-  // compute our current aligned pointer into the shared heap
-  //
-  //   jemalloc 4.5.0 man: "The alignment parameter is always a power of two at
-  //   least as large as the chunk size."
-  cur_chunk_base = alignHelper(heap.base, heap.cur_offset, alignment);
+    cur_heap_size = (uintptr_t)cur_chunk_base - (uintptr_t)heap.base;
 
-  // jemalloc 4.5.0 man: "If chunk is not NULL, the returned pointer must be
-  // chunk on success or NULL on error"
-  if (chunk && chunk != cur_chunk_base) {
-    pthread_mutex_unlock(&heap.alloc_lock);
-    return NULL;
-  }
+    // If there's not enough space on the heap for this allocation, return NULL
+    if (size > heap.size - cur_heap_size) {
+      pthread_mutex_unlock(&heap.alloc_lock);
+      return NULL;
+    }
 
-  cur_heap_size = (uintptr_t)cur_chunk_base - (uintptr_t)heap.base;
+    // Update the current pointer, now that we've past any early returns.
+    heap.cur_offset = cur_heap_size + size;
 
-  // If there's not enough space on the heap for this allocation, return NULL
-  if (size > heap.size - cur_heap_size) {
+    // now that cur_heap_offset is updated, we can unlock
     pthread_mutex_unlock(&heap.alloc_lock);
-    return NULL;
-  }
+  } else {
+    //
+    // Get a dynamic extension chunk.
+    //
+    cur_chunk_base = chpl_comm_regMemAlloc(size);
 
-  // Update the current pointer, now that we've past any early returns.
-  heap.cur_offset = cur_heap_size + size;
+    if (cur_chunk_base == NULL) {
+      return NULL;
+    }
+
+    //
+    // Localize the new memory via first-touch, by storing to each page.
+    // This will give the memory affinity to the NUMA domain (if any)
+    // we were running on when we allocated it.  It also commits the
+    // memory, in the jemalloc sense.
+    //
+    for (int i = 0; i < size; i += heap.page_size) {
+      ((char*) cur_chunk_base)[i] = 0;
```

Comment:
Seems like this should use memset, since memset should be faster (vectorized and maybe other optimizations)

Reply:
Notice the gigantic stride.  We're not actually zeroing memory here.  We're just touching each page to force it into existence and thus localize it.

Reply:
Oooh, duh. Makes sense, nevermind me then. 

I'm fine with this code -- just curious if there's an equivalent hwloc function, and if so if there's any advantages to that approach.

Reply:
There is a hwloc function that will set the locality of memory (that is, pin it).  And there are reasons for and against using it.  The main differences are that touching the memory as we do here sets the locality implicitly (to match the NUMA domain the touching thread executed on) and only sets locality if the page doesn't have any yet, that is, it only has an affect the first time a page is referenced.  But if the page already exists, then it nothing changes.  The hwloc function, on the other hand, accepts an explicit NUMA domain id.  So it can localize the page to a different NUMA domain than we're running on, which might be useful if our expected reference pattern doesn't match that of our parallel initialization code.  Also, it can optionally force the desired locality.  That is, if the page exists and isn't on the desired NUMA domain, it can physically move it to the desired one.  That always gives the desired locality, but does require actually copying the page contents.  So we might want to use the hwloc function if we have memory the parallel init code doesn't work well on, or if we find ourselves reusing memory, but so far neither of those seem to happen much.